clang/test/SemaSYCL/device_global.cpp

```diff
@@ -18,21 +19,80 @@ device_global<char> Foo::d;
 
 struct Baz {
 private:
-  // expected-error@+1{{'device_global' member variable 'f' is not publicly accessible from namespace scope}}
+  // expected-error@+1{{'device_global' member variable 'f' should be publicly accessible from namespace scope}}
   static device_global<int> f;
 };
+
 device_global<int> Baz::f;
 
 device_global<int[4]> not_array; // OK
 
+// expected-error@+1{{'device_global' array is not allowed}}
+device_global<int> array[4];
+
 device_global<int> same_name; // OK
+
 namespace foo {
 device_global<int> same_name; // OK
 }
-namespace {
-device_global<int> same_name; // OK
+
+struct BBar {
+private:
+  struct BarInsider {
+    // expected-error@+1{{'device_global' member variable 'c' should be publicly accessible from namespace scope}}
+    static device_global<float> c;
+  };
+};
+
+struct ABar {
+  void method() {
+    // expected-error@+1{{'device_global' variable must be a static data member or declared in global or namespace scope}}
+    static device_global<float> c;
+  }
+  struct BarInsider {
+    static device_global<float> c;
+    void method() {
+      // expected-error@+1{{'device_global' variable must be a static data member or declared in global or namespace scope}}
+      static device_global<float> c;
+    }
+  };
+};
+
+template <typename T> void fooBar() {
+  // expected-error@+1 2 {{'device_global' variable must be a static data member or declared in global or namespace scope}}
```

Comment:
This is getting diagnosed twice - once at `if (!Var->hasGlobalStorage() || Var->isLocalVarDeclOrParm())` and again at `if (Var->isStaticLocal())` in SemaTemplateInstantiateDecl.cpp